### PR TITLE
Blog app - integration specs for views

### PR DESCRIPTION
**Pull request by Rudy-Menekam**
In this branch, @Zeph180 and I implemented the following

1. Write integration tests with Capybara gem.
2. Secure app from n+1 problems.

Before
Screenshot from 2023-05-23 12-02-12

After
Screenshot from 2023-05-23 18-49-18

### CODE REVIEW

File: spec/views/post_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    expect(page).to have_content(@first_post.title)
+  end
+
+  scenario 'displays post text' do
+    # @first_post.text = 'This is my first post'
+
+    # puts @first_post.text
+    expect(page).to have_content(@first_post.text)
+    # expect(page).to have_css('div.post-header', text: 'This is my first post')
+  end
```
Comment (tresorsawasawa) [inline spec/views/post_index_spec.rb]: - [x] For best practices and clean code, kindly ensure that you remove all the commented and unused code lines from your code base

File: spec/features/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    expect(page).to have_content(post2.title)
+    expect(page).to have_content(post2.text)
+
+    expect(page).to have_link('Create new post', href: new_user_post_path(user))
+    expect(page).to have_link('See all posts', href: "#{user.id}/posts")
+    expect(page).to have_link('Back')
+
+    # it 'it should render the view all user posts link' do
+    #   expect(page).to have_link('See all posts')
+    # end
```
Comment (tresorsawasawa) [inline spec/features/user_show_spec.rb]: - [x]  For best practices and clean code, kindly ensure that you remove all the commented and unused code lines from your code base

File: spec/features/user_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    user1 = User.create(name: 'John')
+    user2 = User.create(name: 'Jane')
+
+    # Visit the view
+    visit users_path
+
+    # Assert that the view contains the user names
+    expect(page).to have_content('List of users')
+    expect(page).to have_content(user1.name)
+    expect(page).to have_content(user2.name)
```
Comment (tresorsawasawa) [inline spec/features/user_index_spec.rb]: - [x] As required for the project, kindly ensure that the `User` **index page** has the following featured tests:
* Tests that asserts you can see the username of all other users.
* Tests that asserts you can see the profile picture for each user.
* Tests that asserts you can see the number of posts each user has written.
* Tests that redirects to the user's show page when you click on a user.

- [ **OPTIONAL** ]**IMPORTANT:** I suggest you put each feature test in `it` method and move both the `User_index_page` and `User_show_page` in the `spec/views` folder for the best practices reasons.

File: spec/features/user_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    expect(page).to have_content(post1.title)
+    expect(page).to have_content(post1.text)
+    expect(page).to have_content(post2.title)
+    expect(page).to have_content(post2.text)
+
+    expect(page).to have_link('Create new post', href: new_user_post_path(user))
+    expect(page).to have_link('See all posts', href: "#{user.id}/posts")
+    expect(page).to have_link('Back')
+
```
Comment (tresorsawasawa) [inline spec/features/user_show_spec.rb]: - [x] As required for the project, kindly ensure that the `User` **show page** has the following featured tests:
   * Tests that asserts you can see the user's profile picture.
   * Tests that asserts you can see the number of posts the user has written.
   * Tests that asserts you can see the user's first 3 posts.
   * Tests that redirects to the post's show page when you click on a user's post.
   * Tests that redirects to the user's post's index page when you click on a **see all posts**.

-- [ **OPTIONAL** ]**IMPORTANT**: I suggest you put each feature test in `it` method and move both the `User_index_page` and `User_show_page` in the spec/views folder for the best practices reasons
```ruby
 it "it should test that if I can see a button that lets me view all of a user's posts." do
    expect(page).to have_link('See all posts')
 end
```

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @Rudy-Menekam and @Zeph180 
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

almost-there

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
- Follow Gitflow :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by omar-muhamad — APPROVED** [review]
# Approved ✔
### Hi Team,

#### Great job, Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

### Congratulations! 🎉 

giphy (1)

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- N/A

## [Optional] suggestions:

- N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**